File: apps/java/src/main/java/chapter10/TribuoClassifier.java
package chapter10;

import chapter02.Features;
import chapter03.TribuoTrees;
import java.util.List;
import org.tribuo.Model;
import org.tribuo.Trainer;
import org.tribuo.classification.Label;
import org.tribuo.classification.dtree.CARTClassificationTrainer;
import org.tribuo.classification.ensemble.VotingCombiner;
import org.tribuo.common.tree.RandomForestTrainer;

/** Tribuo のトレーナーを Classifier に合わせるアダプター。 */
public final class TribuoClassifier implements Classifier {
  /** 深さを制限しないことを表す値 */
  public static final int UNLIMITED = Integer.MAX_VALUE;

  // 分割ごとに、半分の特徴量から分割の候補を選ぶ
  private static final float FRACTION_FEATURES_IN_SPLIT = 0.5f;

  private final Trainer<Label> trainer;
  private Model<Label> model;

  public TribuoClassifier(Trainer<Label> trainer) {
    this.trainer = trainer;
  }

  /** ブートストラップ標本で nEstimators 本の CART を学習し、多数決する Tribuo のランダムフォレスト。 */
  public static TribuoClassifier randomForest(int nEstimators, int maxDepth, long seed) {
    var tree = new CARTClassificationTrainer(maxDepth, FRACTION_FEATURES_IN_SPLIT, seed);
    return new TribuoClassifier(
        new RandomForestTrainer<>(tree, new VotingCombiner(), nEstimators, seed));
  }

  @Override
  public TribuoClassifier fit(List<Features> x, List<String> t) {
    model = trainer.train(TribuoTrees.toDataset(x, t));
    return this;
  }

  @Override
  public List<String> predict(List<Features> x) {
    if (model == null) {
      throw new IllegalStateException("fit で学習してから predict を呼んでください");
    }
    return TribuoTrees.predict(model, x);
  }
}

File: apps/kotlin/src/main/kotlin/chapter10/TribuoClassifier.kt
package chapter10

import chapter03.predictWithTribuo
import chapter03.toTribuoDataset
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.tribuo.Model
import org.tribuo.Trainer
import org.tribuo.classification.Label
import org.tribuo.classification.dtree.CARTClassificationTrainer
import org.tribuo.classification.ensemble.VotingCombiner
import org.tribuo.common.tree.RandomForestTrainer

class TribuoClassifier(
    private val trainer: Trainer<Label>,
) : Classifier {
    private var model: Model<Label>? = null

    override fun fit(
        x: AnyFrame,
        t: List<String>,
    ): TribuoClassifier {
        model = trainer.train(toTribuoDataset(x, t))
        return this
    }

    override fun predict(x: AnyFrame): List<String> {
        val fitted = checkNotNull(model) { "fit で学習してから predict を呼んでください" }
        return predictWithTribuo(fitted, x)
    }
}

// 分割ごとに半分の特徴量から選ぶ決定木を、ブートストラップ標本で nEstimators 本学習して多数決する
private const val FRACTION_FEATURES_IN_SPLIT = 0.5f

fun tribuoRandomForest(
    nEstimators: Int,
    maxDepth: Int?,
    seed: Long,
): TribuoClassifier {
    val tree = CARTClassificationTrainer(maxDepth ?: Int.MAX_VALUE, FRACTION_FEATURES_IN_SPLIT, seed)
    return TribuoClassifier(RandomForestTrainer(tree, VotingCombiner(), nEstimators, seed))
}
